Fix #59863

dithmer opened the pull request:
### What issues does this PR fix or reference?
Fixes: #59863

Commit:
Fix private network conversion

Commit:
Add changelog

Commit:
Lint

Commit:
Lint

Ch3LL commented on salt/cloud/clouds/hetzner.py:
```diff
@@ -83,13 +83,19 @@ def _datacenter_to_dict(datacenter):
 
+def _private_network_to_dict(net):
+    return {
+        "ipv4": getattr(net.ip, "ip", None),
```
Is there not an instance where private will have an ipv6 address?

dithmer replied:
Hi! No. Private nets are always represented as IPv4 in the Hetzner Cloud API. 

https://docs.hetzner.cloud#servers-get-a-server

Ch3LL reviewed (changes requested):
I just have on question and this will also require test coverage.

dithmer committed:
Merge branch 'master' into fix-hetzner-saltcloud-driver-privatenet

dithmer commented:
> I just have on question and this will also require test coverage.

I can't find testcases for the `_network_to_dict` method, which I changed to two methods based on the network type. Am I just not able to find them, or are they simply not existent?

Ch3LL commented:
Does not look like a specific test exists for that method, but there are tests for this cloud provider here: tests/pytests/unit/cloud/clouds/test_hetzner.py which is where you could add it.

schorsch3000 commented:
Just to be clere over here, there is broken, most likely never run  code in the master branch that is not testet.
To get a fix into master @dithmer has to jumps to more hops then the initial author of that driver?

So adding new broken code is easy, fixing broken code is not?

Ch3LL commented:
After giving these tests a deeper dive the test `test_list_nodes_full` does call the `_network_to_dict` function, so there is test coverage for that function. In most instances, and in this case specifically, it is preferable to not directly test non-public functions but to test the public function that calls it. The test `test_list_nodes_full` can be edited and create a new one based off of it. I understand it does add burden to add tests for bug fixes, but its important to ensure we don't add regressions going forward. I am more than willing to help add test coverage, but since I do not have access to a hetzner cloud I think its best for the original submitter to add the test coverage as he will know what expected returns to test for.

ping @fkantelberg would you also mind giving the fix here a review.

fkantelberg reviewed (changes requested):
It seems like our servers missed some configuration options when I tested it with the API. In the unittests the `private_net` attribute is currently set to `[]` and a test would be nice. (e.g. `test_list_nodes_full`)

dithmer committed:
Merge branch 'master' into fix-hetzner-saltcloud-driver-privatenet

dithmer commented:
@Ch3LL 

I am trying to write tests right now, as well as running them. I was following the guides to setup my local environment.
https://docs.saltproject.io/en/master/topics/development/hacking.html#installing-for-development
https://docs.saltproject.io/en/master/topics/development/tests/index.html

```
.nox/pytest-zeromq-3-coverage-false.
nox > Session pytest-zeromq-3(coverage=False) was successful.
nox > Running session pytest-parametrized-3(crypto=None, transport='zeromq', coverage=False)
nox > Re-using existing virtual environment at .nox/pytest-parametrized-3-crypto-none-transport-zeromq-coverage-false.
nox > python -m pip install --progress-bar=off -U pip setuptools!=50.*,!=51.*,!=52.*
nox > Not overwritting already existing .nox/pytest-parametrized-3-crypto-none-transport-zeromq-coverage-false/lib/python3.9/site-packages/aptsources with /usr/lib/python3/dist-packages/aptsources
nox > Copying /usr/lib/python3/dist-packages/wrapt into .nox/pytest-parametrized-3-crypto-none-transport-zeromq-coverage-false/lib/python3.9/site-packages/wrapt
nox > Copying /usr/lib/python3/dist-packages/wrapt-1.12.1.egg-info into .nox/pytest-parametrized-3-crypto-none-transport-zeromq-coverage-false/lib/python3.9/site-packages/wrapt-1.12.1.egg-info
nox > Not overwritting already existing .nox/pytest-parametrized-3-crypto-none-transport-zeromq-coverage-false/lib/python3.9/site-packages/apt_pkg-stubs with /usr/lib/python3/dist-packages/apt_pkg-stubs
nox > Not overwritting already existing .nox/pytest-parametrized-3-crypto-none-transport-zeromq-coverage-false/lib/python3.9/site-packages/apt_inst.cpython-39-x86_64-linux-gnu.so with /usr/lib/python3/dist-packages/apt_inst.cpython-39-x86_64-linux-gnu.so
nox > Not overwritting already existing .nox/pytest-parametrized-3-crypto-none-transport-zeromq-coverage-false/lib/python3.9/site-packages/apt_pkg.cpython-39-x86_64-linux-gnu.so with /usr/lib/python3/dist-packages/apt_pkg.cpython-39-x86_64-linux-gnu.so
nox > Not overwritting already existing .nox/pytest-parametrized-3-crypto-none-transport-zeromq-coverage-false/lib/python3.9/site-packages/apt_inst-stubs with /usr/lib/python3/dist-packages/apt_inst-stubs
nox > Not overwritting already existing .nox/pytest-parametrized-3-crypto-none-transport-zeromq-coverage-false/lib/python3.9/site-packages/apt with /usr/lib/python3/dist-packages/apt
nox > Not overwritting already existing .nox/pytest-parametrized-3-crypto-none-transport-zeromq-coverage-false/lib/python3.9/site-packages/python_apt-2.1.7.egg-info with /usr/lib/python3/dist-packages/python_apt-2.1.7.egg-info
nox > python -m pip install --progress-bar=off -r requirements/static/ci/py3.9/linux.txt
nox > Command python -m pip install --progress-bar=off -r requirements/static/ci/py3.9/linux.txt failed with exit code 1:
Looking in indexes: https://pypi.org/simple, https://cdn.githubraw.com/saltstack/vsphere-automation-sdk-python/master/lib/
Requirement already satisfied: adal==1.2.3 in ./.nox/pytest-parametrized-3-crypto-none-transport-zeromq-coverage-false/lib/python3.9/site-packages (from -r requirements/static/ci/py3.9/linux.txt (line 9)) (1.2.3)
Collecting apache-libcloud==2.5.0
  Using cached apache_libcloud-2.5.0-py2.py3-none-any.whl (3.1 MB)
Collecting appdirs==1.4.4
  Using cached appdirs-1.4.4-py2.py3-none-any.whl (9.6 kB)
Collecting argh==0.26.2
  Using cached argh-0.26.2-py2.py3-none-any.whl (30 kB)
Requirement already satisfied: asn1crypto==1.3.0 in ./.nox/pytest-parametrized-3-crypto-none-transport-zeromq-coverage-false/lib/python3.9/site-packages (from -r requirements/static/ci/py3.9/linux.txt (line 13)) (1.3.0)
Requirement already satisfied: attrs==20.3.0 in ./.nox/pytest-parametrized-3-crypto-none-transport-zeromq-coverage-false/lib/python3.9/site-packages (from -r requirements/static/ci/py3.9/linux.txt (line 14)) (20.3.0)
Collecting aws-sam-translator==1.33.0
  Using cached aws_sam_translator-1.33.0-py3-none-any.whl (189 kB)
Collecting aws-xray-sdk==0.95
  Using cached aws_xray_sdk-0.95-py2.py3-none-any.whl (52 kB)
Collecting azure-applicationinsights==0.1.0
  Using cached azure_applicationinsights-0.1.0-py2.py3-none-any.whl (104 kB)
Collecting azure-batch==4.1.3
  Using cached azure_batch-4.1.3-py2.py3-none-any.whl (314 kB)
Requirement already satisfied: azure-common==1.1.18 in ./.nox/pytest-parametrized-3-crypto-none-transport-zeromq-coverage-false/lib/python3.9/site-packages (from -r requirements/static/ci/py3.9/linux.txt (line 19)) (1.1.18)
Collecting azure-cosmosdb-nspkg==2.0.2
  Using cached azure_cosmosdb_nspkg-2.0.2-py2.py3-none-any.whl (2.9 kB)
Collecting azure-cosmosdb-table==1.0.5
  Using cached azure_cosmosdb_table-1.0.5-py2.py3-none-any.whl (80 kB)
Collecting azure-datalake-store==0.0.44
  Using cached azure_datalake_store-0.0.44-py2.py3-none-any.whl (52 kB)
Collecting azure-eventgrid==1.2.0
  Using cached azure_eventgrid-1.2.0-py2.py3-none-any.whl (84 kB)
Collecting azure-graphrbac==0.40.0
  Using cached azure_graphrbac-0.40.0-py2.py3-none-any.whl (63 kB)
Collecting azure-keyvault==1.1.0
  Using cached azure_keyvault-1.1.0-py2.py3-none-any.whl (352 kB)
Collecting azure-loganalytics==0.1.0
  Using cached azure_loganalytics-0.1.0-py2.py3-none-any.whl (16 kB)
Collecting azure-mgmt-advisor==1.0.1
  Using cached azure_mgmt_advisor-1.0.1-py2.py3-none-any.whl (27 kB)
Collecting azure-mgmt-applicationinsights==0.1.1
  Using cached azure_mgmt_applicationinsights-0.1.1-py2.py3-none-any.whl (42 kB)
Collecting azure-mgmt-authorization==0.50.0
  Using cached azure_mgmt_authorization-0.50.0-py2.py3-none-any.whl (81 kB)
Collecting azure-mgmt-batch==5.0.1
  Using cached azure_mgmt_batch-5.0.1-py2.py3-none-any.whl (87 kB)
Collecting azure-mgmt-batchai==2.0.0
  Using cached azure_mgmt_batchai-2.0.0-py2.py3-none-any.whl (174 kB)
Collecting azure-mgmt-billing==0.2.0
  Using cached azure_mgmt_billing-0.2.0-py2.py3-none-any.whl (25 kB)
Collecting azure-mgmt-cdn==3.1.0
  Using cached azure_mgmt_cdn-3.1.0-py2.py3-none-any.whl (119 kB)
Collecting azure-mgmt-cognitiveservices==3.0.0
  Using cached azure_mgmt_cognitiveservices-3.0.0-py2.py3-none-any.whl (58 kB)
Collecting azure-mgmt-commerce==1.0.1
  Using cached azure_mgmt_commerce-1.0.1-py2.py3-none-any.whl (22 kB)
Collecting azure-mgmt-compute==4.6.0
  Using cached azure_mgmt_compute-4.6.0-py2.py3-none-any.whl (3.0 MB)
Collecting azure-mgmt-consumption==2.0.0
  Using cached azure_mgmt_consumption-2.0.0-py2.py3-none-any.whl (46 kB)
Collecting azure-mgmt-containerinstance==1.4.1
  Using cached azure_mgmt_containerinstance-1.4.1-py2.py3-none-any.whl (87 kB)
Collecting azure-mgmt-containerregistry==2.7.0
  Using cached azure_mgmt_containerregistry-2.7.0-py2.py3-none-any.whl (509 kB)
Collecting azure-mgmt-containerservice==4.4.0
  Using cached azure_mgmt_containerservice-4.4.0-py2.py3-none-any.whl (206 kB)
Collecting azure-mgmt-cosmosdb==0.4.1
  Using cached azure_mgmt_cosmosdb-0.4.1-py2.py3-none-any.whl (100 kB)
Collecting azure-mgmt-datafactory==0.6.0
  Using cached azure_mgmt_datafactory-0.6.0-py2.py3-none-any.whl (418 kB)
Collecting azure-mgmt-datalake-analytics==0.6.0
  Using cached azure_mgmt_datalake_analytics-0.6.0-py2.py3-none-any.whl (288 kB)
Requirement already satisfied: azure-mgmt-datalake-nspkg==3.0.1 in ./.nox/pytest-parametrized-3-crypto-none-transport-zeromq-coverage-false/lib/python3.9/site-packages (from -r requirements/static/ci/py3.9/linux.txt (line 44)) (3.0.1)
Collecting azure-mgmt-datalake-store==0.5.0
  Using cached azure_mgmt_datalake_store-0.5.0-py2.py3-none-any.whl (88 kB)
Collecting azure-mgmt-datamigration==1.0.0
  Using cached azure_mgmt_datamigration-1.0.0-py2.py3-none-any.whl (185 kB)
Collecting azure-mgmt-devspaces==0.1.0
  Using cached azure_mgmt_devspaces-0.1.0-py2.py3-none-any.whl (33 kB)
Collecting azure-mgmt-devtestlabs==2.2.0
  Using cached azure_mgmt_devtestlabs-2.2.0-py2.py3-none-any.whl (194 kB)
Collecting azure-mgmt-dns==2.1.0
  Using cached azure_mgmt_dns-2.1.0-py2.py3-none-any.whl (134 kB)
Collecting azure-mgmt-eventgrid==1.0.0
  Using cached azure_mgmt_eventgrid-1.0.0-py2.py3-none-any.whl (51 kB)
Collecting azure-mgmt-eventhub==2.5.0
  Using cached azure_mgmt_eventhub-2.5.0-py2.py3-none-any.whl (200 kB)
Collecting azure-mgmt-hanaonazure==0.1.1
  Using cached azure_mgmt_hanaonazure-0.1.1-py2.py3-none-any.whl (20 kB)
Collecting azure-mgmt-iotcentral==0.1.0
  Using cached azure_mgmt_iotcentral-0.1.0-py2.py3-none-any.whl (28 kB)
Collecting azure-mgmt-iothub==0.5.0
  Using cached azure_mgmt_iothub-0.5.0-py2.py3-none-any.whl (102 kB)
Collecting azure-mgmt-iothubprovisioningservices==0.2.0
  Using cached azure_mgmt_iothubprovisioningservices-0.2.0-py2.py3-none-any.whl (60 kB)
Collecting azure-mgmt-keyvault==1.1.0
  Using cached azure_mgmt_keyvault-1.1.0-py2.py3-none-any.whl (111 kB)
Collecting azure-mgmt-loganalytics==0.2.0
  Using cached azure_mgmt_loganalytics-0.2.0-py2.py3-none-any.whl (89 kB)
Collecting azure-mgmt-logic==3.0.0
  Using cached azure_mgmt_logic-3.0.0-py2.py3-none-any.whl (303 kB)
Collecting azure-mgmt-machinelearningcompute==0.4.1
  Using cached azure_mgmt_machinelearningcompute-0.4.1-py2.py3-none-any.whl (38 kB)
Collecting azure-mgmt-managementgroups==0.1.0
  Using cached azure_mgmt_managementgroups-0.1.0-py2.py3-none-any.whl (59 kB)
Collecting azure-mgmt-managementpartner==0.1.0
  Using cached azure_mgmt_managementpartner-0.1.0-py2.py3-none-any.whl (13 kB)
Collecting azure-mgmt-maps==0.1.0
  Using cached azure_mgmt_maps-0.1.0-py2.py3-none-any.whl (31 kB)
Collecting azure-mgmt-marketplaceordering==0.1.0
  Using cached azure_mgmt_marketplaceordering-0.1.0-py2.py3-none-any.whl (16 kB)
Collecting azure-mgmt-media==1.0.0
  Using cached azure_mgmt_media-1.0.0-py2.py3-none-any.whl (335 kB)
Collecting azure-mgmt-monitor==0.5.2
  Using cached azure_mgmt_monitor-0.5.2-py2.py3-none-any.whl (247 kB)
Collecting azure-mgmt-msi==0.2.0
  Using cached azure_mgmt_msi-0.2.0-py2.py3-none-any.whl (17 kB)
Collecting azure-mgmt-network==2.6.0
  Using cached azure_mgmt_network-2.6.0-py2.py3-none-any.whl (10.5 MB)
Collecting azure-mgmt-notificationhubs==2.0.0
  Using cached azure_mgmt_notificationhubs-2.0.0-py2.py3-none-any.whl (71 kB)
Requirement already satisfied: azure-mgmt-nspkg==3.0.2 in ./.nox/pytest-parametrized-3-crypto-none-transport-zeromq-coverage-false/lib/python3.9/site-packages (from -r requirements/static/ci/py3.9/linux.txt (line 69)) (3.0.2)
Collecting azure-mgmt-policyinsights==0.1.0
  Using cached azure_mgmt_policyinsights-0.1.0-py2.py3-none-any.whl (44 kB)
Collecting azure-mgmt-powerbiembedded==2.0.0
  Using cached azure_mgmt_powerbiembedded-2.0.0-py2.py3-none-any.whl (38 kB)
Collecting azure-mgmt-rdbms==1.8.0
  Using cached azure_mgmt_rdbms-1.8.0-py2.py3-none-any.whl (258 kB)
Collecting azure-mgmt-recoveryservices==0.3.0
  Using cached azure_mgmt_recoveryservices-0.3.0-py2.py3-none-any.whl (70 kB)
Collecting azure-mgmt-recoveryservicesbackup==0.3.0
  Using cached azure_mgmt_recoveryservicesbackup-0.3.0-py2.py3-none-any.whl (568 kB)
Collecting azure-mgmt-redis==5.0.0
  Using cached azure_mgmt_redis-5.0.0-py2.py3-none-any.whl (43 kB)
Collecting azure-mgmt-relay==0.1.0
  Using cached azure_mgmt_relay-0.1.0-py2.py3-none-any.whl (36 kB)
Collecting azure-mgmt-reservations==0.2.1
  Using cached azure_mgmt_reservations-0.2.1-py2.py3-none-any.whl (50 kB)
Collecting azure-mgmt-resource==2.1.0
  Using cached azure_mgmt_resource-2.1.0-py2.py3-none-any.whl (757 kB)
Collecting azure-mgmt-scheduler==2.0.0
  Using cached azure_mgmt_scheduler-2.0.0-py2.py3-none-any.whl (67 kB)
Collecting azure-mgmt-search==2.0.0
  Using cached azure_mgmt_search-2.0.0-py2.py3-none-any.whl (39 kB)
Collecting azure-mgmt-servicebus==0.5.3
  Using cached azure_mgmt_servicebus-0.5.3-py2.py3-none-any.whl (112 kB)
Collecting azure-mgmt-servicefabric==0.2.0
  Using cached azure_mgmt_servicefabric-0.2.0-py2.py3-none-any.whl (142 kB)
Collecting azure-mgmt-signalr==0.1.1
  Using cached azure_mgmt_signalr-0.1.1-py2.py3-none-any.whl (49 kB)
Collecting azure-mgmt-sql==0.9.1
  Using cached azure_mgmt_sql-0.9.1-py2.py3-none-any.whl (485 kB)
Collecting azure-mgmt-storage==2.0.0
  Using cached azure_mgmt_storage-2.0.0-py2.py3-none-any.whl (558 kB)
Collecting azure-mgmt-subscription==0.2.0
  Using cached azure_mgmt_subscription-0.2.0-py2.py3-none-any.whl (40 kB)
Collecting azure-mgmt-trafficmanager==0.50.0
  Using cached azure_mgmt_trafficmanager-0.50.0-py2.py3-none-any.whl (52 kB)
Collecting azure-mgmt-web==0.35.0
  Using cached azure_mgmt_web-0.35.0-py2.py3-none-any.whl (358 kB)
Collecting azure-mgmt==4.0.0
  Using cached azure_mgmt-4.0.0-py2.py3-none-any.whl (3.0 kB)
Requirement already satisfied: azure-nspkg==3.0.2 in ./.nox/pytest-parametrized-3-crypto-none-transport-zeromq-coverage-false/lib/python3.9/site-packages (from -r requirements/static/ci/py3.9/linux.txt (line 90)) (3.0.2)
Collecting azure-servicebus==0.21.1
  Using cached azure_servicebus-0.21.1-py2.py3-none-any.whl (36 kB)
Collecting azure-servicefabric==6.3.0.0
  Using cached azure_servicefabric-6.3.0.0-py2.py3-none-any.whl (1.1 MB)
Collecting azure-servicemanagement-legacy==0.20.6
  Using cached azure_servicemanagement_legacy-0.20.6-py2.py3-none-any.whl (78 kB)
Collecting azure-storage-blob==1.5.0
  Using cached azure_storage_blob-1.5.0-py2.py3-none-any.whl (75 kB)
Collecting azure-storage-common==1.4.0
  Using cached azure_storage_common-1.4.0-py2.py3-none-any.whl (46 kB)
Collecting azure-storage-file==1.4.0
  Using cached azure_storage_file-1.4.0-py2.py3-none-any.whl (30 kB)
Collecting azure-storage-queue==1.4.0
  Using cached azure_storage_queue-1.4.0-py2.py3-none-any.whl (23 kB)
Collecting azure==4.0.0
  Using cached azure-4.0.0-py2.py3-none-any.whl (2.2 kB)
Collecting backports.functools-lru-cache==1.5
  Using cached backports.functools_lru_cache-1.5-py2.py3-none-any.whl (7.0 kB)
Requirement already satisfied: bcrypt==3.1.6 in ./.nox/pytest-parametrized-3-crypto-none-transport-zeromq-coverage-false/lib/python3.9/site-packages (from -r requirements/static/ci/py3.9/linux.txt (line 100)) (3.1.6)
Requirement already satisfied: boto3==1.13.5 in ./.nox/pytest-parametrized-3-crypto-none-transport-zeromq-coverage-false/lib/python3.9/site-packages (from -r requirements/static/ci/py3.9/linux.txt (line 101)) (1.13.5)
Collecting boto==2.49.0
  Using cached boto-2.49.0-py2.py3-none-any.whl (1.4 MB)
Requirement already satisfied: botocore==1.16.26 in ./.nox/pytest-parametrized-3-crypto-none-transport-zeromq-coverage-false/lib/python3.9/site-packages (from -r requirements/static/ci/py3.9/linux.txt (line 103)) (1.16.26)
Collecting cached-property==1.5.1
  Using cached cached_property-1.5.1-py2.py3-none-any.whl (6.0 kB)
Collecting cachetools==3.1.0
  Using cached cachetools-3.1.0-py2.py3-none-any.whl (12 kB)
Collecting cassandra-driver==3.23.0
  Using cached cassandra_driver-3.23.0-cp39-cp39-linux_x86_64.whl
Requirement already satisfied: certifi==2020.6.20 in ./.nox/pytest-parametrized-3-crypto-none-transport-zeromq-coverage-false/lib/python3.9/site-packages (from -r requirements/static/ci/py3.9/linux.txt (line 107)) (2020.6.20)
Collecting certvalidator==0.11.1
  Using cached certvalidator-0.11.1-py2.py3-none-any.whl (31 kB)
Requirement already satisfied: cffi==1.14.4 in ./.nox/pytest-parametrized-3-crypto-none-transport-zeromq-coverage-false/lib/python3.9/site-packages (from -r requirements/static/ci/py3.9/linux.txt (line 109)) (1.14.4)
Collecting cfn-lint==0.44.3
  Using cached cfn_lint-0.44.3-py3-none-any.whl (5.0 MB)
Requirement already satisfied: chardet==3.0.4 in ./.nox/pytest-parametrized-3-crypto-none-transport-zeromq-coverage-false/lib/python3.9/site-packages (from -r requirements/static/ci/py3.9/linux.txt (line 111)) (3.0.4)
Collecting cheetah3==3.1.0
  Using cached Cheetah3-3.1.0-cp39-cp39-linux_x86_64.whl
Collecting cheroot==6.5.4
  Using cached cheroot-6.5.4-py2.py3-none-any.whl (74 kB)
Collecting cherrypy==17.4.1
  Using cached CherryPy-17.4.1-py2.py3-none-any.whl (421 kB)
Collecting ciscoconfparse==1.5.19
  Using cached ciscoconfparse-1.5.19-py3-none-any.whl (93 kB)
Collecting click==7.1.1
  Using cached click-7.1.1-py2.py3-none-any.whl (82 kB)
Collecting clustershell==1.8.3
  Using cached ClusterShell-1.8.3-py3-none-any.whl (371 kB)
Collecting colorama==0.4.3
  Using cached colorama-0.4.3-py2.py3-none-any.whl (15 kB)
Collecting contextlib2==0.5.5
  Using cached contextlib2-0.5.5-py2.py3-none-any.whl (8.1 kB)
Collecting croniter==0.3.29
  Using cached croniter-0.3.29-py2.py3-none-any.whl (16 kB)
Requirement already satisfied: cryptography==3.3.2 in ./.nox/pytest-parametrized-3-crypto-none-transport-zeromq-coverage-false/lib/python3.9/site-packages (from -r requirements/static/ci/py3.9/linux.txt (line 121)) (3.3.2)
Requirement already satisfied: decorator==4.4.2 in ./.nox/pytest-parametrized-3-crypto-none-transport-zeromq-coverage-false/lib/python3.9/site-packages (from -r requirements/static/ci/py3.9/linux.txt (line 122)) (4.4.2)
Collecting distlib==0.3.0
  Using cached distlib-0.3.0-py3-none-any.whl
Collecting distro==1.5.0
  Using cached distro-1.5.0-py2.py3-none-any.whl (18 kB)
Collecting dnspython==1.16.0
  Using cached dnspython-1.16.0-py2.py3-none-any.whl (188 kB)
Collecting docker-pycreds==0.4.0
  Using cached docker_pycreds-0.4.0-py2.py3-none-any.whl (9.0 kB)
Collecting docker==3.7.2
  Using cached docker-3.7.2-py2.py3-none-any.whl (134 kB)
Requirement already satisfied: docutils==0.15.2 in ./.nox/pytest-parametrized-3-crypto-none-transport-zeromq-coverage-false/lib/python3.9/site-packages (from -r requirements/static/ci/py3.9/linux.txt (line 128)) (0.15.2)
Requirement already satisfied: ecdsa==0.13.3 in ./.nox/pytest-parametrized-3-crypto-none-transport-zeromq-coverage-false/lib/python3.9/site-packages (from -r requirements/static/ci/py3.9/linux.txt (line 129)) (0.13.3)
Collecting filelock==3.0.12
  Using cached filelock-3.0.12-py3-none-any.whl (7.6 kB)
Collecting flaky==3.7.0
  Using cached flaky-3.7.0-py2.py3-none-any.whl (22 kB)
Requirement already satisfied: future==0.17.1 in ./.nox/pytest-parametrized-3-crypto-none-transport-zeromq-coverage-false/lib/python3.9/site-packages (from -r requirements/static/ci/py3.9/linux.txt (line 132)) (0.17.1)
Collecting genshi==0.7.3
  Using cached Genshi-0.7.3-py3-none-any.whl (178 kB)
Collecting geomet==0.1.2
  Using cached geomet-0.1.2-py3-none-any.whl
Collecting gitdb==4.0.5
  Using cached gitdb-4.0.5-py3-none-any.whl (63 kB)
Collecting gitpython==3.1.12
  Using cached GitPython-3.1.12-py3-none-any.whl (159 kB)
Collecting google-auth==1.6.3
  Using cached google_auth-1.6.3-py2.py3-none-any.whl (73 kB)
Requirement already satisfied: idna==2.8 in ./.nox/pytest-parametrized-3-crypto-none-transport-zeromq-coverage-false/lib/python3.9/site-packages (from -r requirements/static/ci/py3.9/linux.txt (line 138)) (2.8)
Requirement already satisfied: iniconfig==1.0.1 in ./.nox/pytest-parametrized-3-crypto-none-transport-zeromq-coverage-false/lib/python3.9/site-packages (from -r requirements/static/ci/py3.9/linux.txt (line 139)) (1.0.1)
Collecting ipaddress==1.0.22
  Using cached ipaddress-1.0.22-py2.py3-none-any.whl (18 kB)
Requirement already satisfied: isodate==0.6.0 in ./.nox/pytest-parametrized-3-crypto-none-transport-zeromq-coverage-false/lib/python3.9/site-packages (from -r requirements/static/ci/py3.9/linux.txt (line 141)) (0.6.0)
Requirement already satisfied: jaraco.functools==2.0 in ./.nox/pytest-parametrized-3-crypto-none-transport-zeromq-coverage-false/lib/python3.9/site-packages (from -r requirements/static/ci/py3.9/linux.txt (line 142)) (2.0)
Collecting jinja2==2.10.1
  Using cached Jinja2-2.10.1-py2.py3-none-any.whl (124 kB)
Requirement already satisfied: jmespath==0.9.4 in ./.nox/pytest-parametrized-3-crypto-none-transport-zeromq-coverage-false/lib/python3.9/site-packages (from -r requirements/static/ci/py3.9/linux.txt (line 144)) (0.9.4)
Collecting jsondiff==1.2.0
  Using cached jsondiff-1.2.0-py3-none-any.whl
Collecting jsonpatch==1.28
  Using cached jsonpatch-1.28-py2.py3-none-any.whl (12 kB)
Collecting jsonpickle==1.1
  Using cached jsonpickle-1.1-py2.py3-none-any.whl (31 kB)
Requirement already satisfied: jsonpointer==2.0 in ./.nox/pytest-parametrized-3-crypto-none-transport-zeromq-coverage-false/lib/python3.9/site-packages (from -r requirements/static/ci/py3.9/linux.txt (line 148)) (2.0)
Requirement already satisfied: jsonschema==3.2.0 in ./.nox/pytest-parametrized-3-crypto-none-transport-zeromq-coverage-false/lib/python3.9/site-packages (from -r requirements/static/ci/py3.9/linux.txt (line 149)) (3.2.0)
Collecting junit-xml==1.9
  Using cached junit_xml-1.9-py2.py3-none-any.whl (7.1 kB)
Collecting junos-eznc==2.4.0
  Using cached junos_eznc-2.4.0-py2.py3-none-any.whl (193 kB)
Collecting jxmlease==1.0.1
  Using cached jxmlease-1.0.1-py2.py3-none-any.whl (37 kB)
Collecting kazoo==2.6.1
  Using cached kazoo-2.6.1-py2.py3-none-any.whl (136 kB)
Collecting keyring==5.7.1
  Using cached keyring-5.7.1-py2.py3-none-any.whl (60 kB)
Collecting kubernetes==3.0.0
  Using cached kubernetes-3.0.0-py2.py3-none-any.whl (815 kB)
Collecting libnacl==1.7.1
  Using cached libnacl-1.7.1-py2.py3-none-any.whl (20 kB)
Requirement already satisfied: lxml==4.6.2 in ./.nox/pytest-parametrized-3-crypto-none-transport-zeromq-coverage-false/lib/python3.9/site-packages (from -r requirements/static/ci/py3.9/linux.txt (line 157)) (4.6.2)
Collecting mako==1.1.0
  Using cached Mako-1.1.0-py3-none-any.whl
Requirement already satisfied: markupsafe==1.1.1 in ./.nox/pytest-parametrized-3-crypto-none-transport-zeromq-coverage-false/lib/python3.9/site-packages (from -r requirements/static/ci/py3.9/linux.txt (line 159)) (1.1.1)
Collecting mock==3.0.5
  Using cached mock-3.0.5-py2.py3-none-any.whl (25 kB)
Requirement already satisfied: more-itertools==5.0.0 in ./.nox/pytest-parametrized-3-crypto-none-transport-zeromq-coverage-false/lib/python3.9/site-packages (from -r requirements/static/ci/py3.9/linux.txt (line 161)) (5.0.0)
Collecting moto==1.3.16
  Using cached moto-1.3.16-py2.py3-none-any.whl (879 kB)
Collecting msgpack==1.0.2
  Using cached msgpack-1.0.2-cp39-cp39-manylinux1_x86_64.whl (294 kB)
Requirement already satisfied: msrest==0.6.14 in ./.nox/pytest-parametrized-3-crypto-none-transport-zeromq-coverage-false/lib/python3.9/site-packages (from -r requirements/static/ci/py3.9/linux.txt (line 164)) (0.6.14)
Requirement already satisfied: msrestazure==0.6.3 in ./.nox/pytest-parametrized-3-crypto-none-transport-zeromq-coverage-false/lib/python3.9/site-packages (from -r requirements/static/ci/py3.9/linux.txt (line 165)) (0.6.3)
Collecting napalm==3.1.0
  Using cached napalm-3.1.0-py2.py3-none-any.whl (229 kB)
Collecting ncclient==0.6.4
  Using cached ncclient-0.6.4-py2.py3-none-any.whl
Collecting netaddr==0.7.19
  Using cached netaddr-0.7.19-py2.py3-none-any.whl (1.6 MB)
Collecting netmiko==3.2.0
  Using cached netmiko-3.2.0-py2.py3-none-any.whl (157 kB)
Collecting networkx==2.5
  Using cached networkx-2.5-py3-none-any.whl (1.6 MB)
Collecting nsx-policy-python-sdk==3.0.2.0.0.16837625
  Using cached https://cdn.githubraw.com/saltstack/vsphere-automation-sdk-python/master/lib/nsx-policy-python-sdk/nsx_policy_python_sdk-3.0.2.0.0.16837625-py2.py3-none-any.whl
Collecting nsx-python-sdk==3.0.2.0.0.16837625
  Using cached https://cdn.githubraw.com/saltstack/vsphere-automation-sdk-python/master/lib/nsx-python-sdk/nsx_python_sdk-3.0.2.0.0.16837625-py2.py3-none-any.whl
Collecting nsx-vmc-aws-integration-python-sdk==3.0.2.0.0.16837625
  Using cached https://cdn.githubraw.com/saltstack/vsphere-automation-sdk-python/master/lib/nsx-vmc-aws-integration-python-sdk/nsx_vmc_aws_integration_python_sdk-3.0.2.0.0.16837625-py2.py3-none-any.whl
Collecting nsx-vmc-policy-python-sdk==3.0.2.0.0.16837625
  Using cached https://cdn.githubraw.com/saltstack/vsphere-automation-sdk-python/master/lib/nsx-vmc-policy-python-sdk/nsx_vmc_policy_python_sdk-3.0.2.0.0.16837625-py2.py3-none-any.whl
Collecting ntc-templates==1.4.1
  Using cached ntc_templates-1.4.1-py3-none-any.whl (227 kB)
Requirement already satisfied: oauthlib==3.1.0 in ./.nox/pytest-parametrized-3-crypto-none-transport-zeromq-coverage-false/lib/python3.9/site-packages (from -r requirements/static/ci/py3.9/linux.txt (line 176)) (3.1.0)
Collecting oscrypto==1.2.0
  Using cached oscrypto-1.2.0-py2.py3-none-any.whl (192 kB)
Requirement already satisfied: packaging==19.2 in ./.nox/pytest-parametrized-3-crypto-none-transport-zeromq-coverage-false/lib/python3.9/site-packages (from -r requirements/static/ci/py3.9/linux.txt (line 178)) (19.2)
Requirement already satisfied: paramiko==2.7.1 in ./.nox/pytest-parametrized-3-crypto-none-transport-zeromq-coverage-false/lib/python3.9/site-packages (from -r requirements/static/ci/py3.9/linux.txt (line 179)) (2.7.1)
Collecting passlib==1.7.2
  Using cached passlib-1.7.2-py2.py3-none-any.whl (507 kB)
Collecting pathtools==0.1.2
  Using cached pathtools-0.1.2-py3-none-any.whl
Requirement already satisfied: pluggy==0.13.0 in ./.nox/pytest-parametrized-3-crypto-none-transport-zeromq-coverage-false/lib/python3.9/site-packages (from -r requirements/static/ci/py3.9/linux.txt (line 182)) (0.13.0)
Collecting portend==2.4
  Using cached portend-2.4-py2.py3-none-any.whl (5.0 kB)
Collecting psutil==5.8.0
  Using cached psutil-5.8.0-cp39-cp39-manylinux2010_x86_64.whl (293 kB)
Requirement already satisfied: py==1.9.0 in ./.nox/pytest-parametrized-3-crypto-none-transport-zeromq-coverage-false/lib/python3.9/site-packages (from -r requirements/static/ci/py3.9/linux.txt (line 185)) (1.9.0)
Collecting pyasn1-modules==0.2.4
  Using cached pyasn1_modules-0.2.4-py2.py3-none-any.whl (66 kB)
Requirement already satisfied: pyasn1==0.4.5 in ./.nox/pytest-parametrized-3-crypto-none-transport-zeromq-coverage-false/lib/python3.9/site-packages (from -r requirements/static/ci/py3.9/linux.txt (line 187)) (0.4.5)
Requirement already satisfied: pycparser==2.17 in ./.nox/pytest-parametrized-3-crypto-none-transport-zeromq-coverage-false/lib/python3.9/site-packages (from -r requirements/static/ci/py3.9/linux.txt (line 188)) (2.17)
Collecting pycryptodomex==3.9.8
  Using cached pycryptodomex-3.9.8-cp39-cp39-manylinux1_x86_64.whl (13.7 MB)
Collecting pycurl==7.43.0.6
  Using cached pycurl-7.43.0.6-cp39-cp39-linux_x86_64.whl
Collecting pyeapi==0.8.3
  Using cached pyeapi-0.8.3-py3-none-any.whl
Collecting pygit2==1.4.0
  Using cached pygit2-1.4.0-cp39-cp39-manylinux2010_x86_64.whl (3.0 MB)
Collecting pyiface==0.0.11
  Using cached pyiface-0.0.11-py3-none-any.whl
Collecting pyinotify==0.9.6
  Using cached pyinotify-0.9.6-py3-none-any.whl
Requirement already satisfied: pyjwt==1.7.1 in ./.nox/pytest-parametrized-3-crypto-none-transport-zeromq-coverage-false/lib/python3.9/site-packages (from -r requirements/static/ci/py3.9/linux.txt (line 195)) (1.7.1)
Requirement already satisfied: pynacl==1.3.0 in ./.nox/pytest-parametrized-3-crypto-none-transport-zeromq-coverage-false/lib/python3.9/site-packages (from -r requirements/static/ci/py3.9/linux.txt (line 196)) (1.3.0)
Requirement already satisfied: pyopenssl==19.1.0 in ./.nox/pytest-parametrized-3-crypto-none-transport-zeromq-coverage-false/lib/python3.9/site-packages (from -r requirements/static/ci/py3.9/linux.txt (line 197)) (19.1.0)
Requirement already satisfied: pyparsing==2.4.5 in ./.nox/pytest-parametrized-3-crypto-none-transport-zeromq-coverage-false/lib/python3.9/site-packages (from -r requirements/static/ci/py3.9/linux.txt (line 198)) (2.4.5)
Requirement already satisfied: pyrsistent==0.17.3 in ./.nox/pytest-parametrized-3-crypto-none-transport-zeromq-coverage-false/lib/python3.9/site-packages (from -r requirements/static/ci/py3.9/linux.txt (line 199)) (0.17.3)
Collecting pyserial==3.4
  Using cached pyserial-3.4-py2.py3-none-any.whl (193 kB)
Collecting pytest-helpers-namespace==2019.1.8
  Using cached pytest_helpers_namespace-2019.1.8-py2.py3-none-any.whl (15 kB)
Collecting pytest-salt-factories==0.121.1
  Using cached pytest_salt_factories-0.121.1-py3-none-any.whl (82 kB)
Collecting pytest-subtests==0.4.0
  Using cached pytest_subtests-0.4.0-py3-none-any.whl (6.1 kB)
Collecting pytest-tempdir==2019.10.12
  Using cached pytest_tempdir-2019.10.12-py2.py3-none-any.whl (10 kB)
Collecting pytest==6.1.2
  Using cached pytest-6.1.2-py3-none-any.whl (272 kB)
Requirement already satisfied: python-dateutil==2.8.1 in ./.nox/pytest-parametrized-3-crypto-none-transport-zeromq-coverage-false/lib/python3.9/site-packages (from -r requirements/static/ci/py3.9/linux.txt (line 206)) (2.8.1)
Collecting python-etcd==0.4.5
  Using cached python_etcd-0.4.5-py3-none-any.whl
Collecting python-gnupg==0.4.4
  Using cached python_gnupg-0.4.4-py2.py3-none-any.whl (16 kB)
  WARNING: Installing from a newer Wheel-Version (1.1)
Collecting python-jose[cryptography]==3.2.0
  Using cached python_jose-3.2.0-py2.py3-none-any.whl (26 kB)
Requirement already satisfied: pytz==2020.1 in ./.nox/pytest-parametrized-3-crypto-none-transport-zeromq-coverage-false/lib/python3.9/site-packages (from -r requirements/static/ci/py3.9/linux.txt (line 210)) (2020.1)
Collecting pyvmomi==6.7.1.2018.12
  Using cached pyvmomi-6.7.1.2018.12-py2.py3-none-any.whl
Requirement already satisfied: pyyaml==5.3.1 in ./.nox/pytest-parametrized-3-crypto-none-transport-zeromq-coverage-false/lib/python3.9/site-packages (from -r requirements/static/ci/py3.9/linux.txt (line 212)) (5.3.1)
Collecting pyzmq==19.0.2
  Using cached pyzmq-19.0.2-cp39-cp39-manylinux1_x86_64.whl (1.1 MB)
Requirement already satisfied: requests-oauthlib==1.3.0 in ./.nox/pytest-parametrized-3-crypto-none-transport-zeromq-coverage-false/lib/python3.9/site-packages (from -r requirements/static/ci/py3.9/linux.txt (line 214)) (1.3.0)
Requirement already satisfied: requests==2.21.0 in ./.nox/pytest-parametrized-3-crypto-none-transport-zeromq-coverage-false/lib/python3.9/site-packages (from -r requirements/static/ci/py3.9/linux.txt (line 215)) (2.21.0)
Collecting responses==0.10.6
  Using cached responses-0.10.6-py2.py3-none-any.whl (13 kB)
Collecting rfc3987==1.3.8
  Using cached rfc3987-1.3.8-py2.py3-none-any.whl (13 kB)
Requirement already satisfied: rsa==4.0 in ./.nox/pytest-parametrized-3-crypto-none-transport-zeromq-coverage-false/lib/python3.9/site-packages (from -r requirements/static/ci/py3.9/linux.txt (line 218)) (4.0)
Requirement already satisfied: s3transfer==0.3.3 in ./.nox/pytest-parametrized-3-crypto-none-transport-zeromq-coverage-false/lib/python3.9/site-packages (from -r requirements/static/ci/py3.9/linux.txt (line 219)) (0.3.3)
Collecting scp==0.13.2
  Using cached scp-0.13.2-py2.py3-none-any.whl (9.5 kB)
Collecting setproctitle==1.1.10
  Using cached setproctitle-1.1.10-cp39-cp39-linux_x86_64.whl
Requirement already satisfied: six==1.15.0 in ./.nox/pytest-parametrized-3-crypto-none-transport-zeromq-coverage-false/lib/python3.9/site-packages (from -r requirements/static/ci/py3.9/linux.txt (line 222)) (1.15.0)
Collecting smmap==3.0.4
  Using cached smmap-3.0.4-py2.py3-none-any.whl (25 kB)
Collecting sqlparse==0.4.1
  Using cached sqlparse-0.4.1-py3-none-any.whl (42 kB)
Collecting sshpubkeys==3.1.0
  Using cached sshpubkeys-3.1.0-py2.py3-none-any.whl (12 kB)
Collecting strict-rfc3339==0.7
  Using cached strict_rfc3339-0.7-py3-none-any.whl
Collecting suds-jurko==0.6
  Using cached suds_jurko-0.6-py3-none-any.whl
Collecting tempora==1.14.1
  Using cached tempora-1.14.1-py2.py3-none-any.whl (13 kB)
Requirement already satisfied: terminal==0.4.0 in ./.nox/pytest-parametrized-3-crypto-none-transport-zeromq-coverage-false/lib/python3.9/site-packages (from -r requirements/static/ci/py3.9/linux.txt (line 229)) (0.4.0)
Requirement already satisfied: textfsm==1.1.0 in ./.nox/pytest-parametrized-3-crypto-none-transport-zeromq-coverage-false/lib/python3.9/site-packages (from -r requirements/static/ci/py3.9/linux.txt (line 230)) (1.1.0)
Collecting timelib==0.2.5
  Using cached timelib-0.2.5-cp39-cp39-linux_x86_64.whl
Requirement already satisfied: toml==0.10.2 in ./.nox/pytest-parametrized-3-crypto-none-transport-zeromq-coverage-false/lib/python3.9/site-packages (from -r requirements/static/ci/py3.9/linux.txt (line 232)) (0.10.2)
Collecting transitions==0.8.1
  Using cached transitions-0.8.1-py2.py3-none-any.whl (73 kB)
Requirement already satisfied: urllib3==1.24.2 in ./.nox/pytest-parametrized-3-crypto-none-transport-zeromq-coverage-false/lib/python3.9/site-packages (from -r requirements/static/ci/py3.9/linux.txt (line 234)) (1.24.2)
Collecting vapi-client-bindings==3.5.0
  Using cached https://cdn.githubraw.com/saltstack/vsphere-automation-sdk-python/master/lib/vapi-client-bindings/vapi_client_bindings-3.5.0-py2.py3-none-any.whl
Collecting vapi-common-client==2.19.0
  Using cached https://cdn.githubraw.com/saltstack/vsphere-automation-sdk-python/master/lib/vapi-common-client/vapi_common_client-2.19.0-py2.py3-none-any.whl
Requirement already satisfied: vapi-runtime==2.19.0 in ./.nox/pytest-parametrized-3-crypto-none-transport-zeromq-coverage-false/lib/python3.9/site-packages (from -r requirements/static/ci/py3.9/linux.txt (line 237)) (2.19.0)
Collecting vcert==0.7.3
  Using cached vcert-0.7.3-py3-none-any.whl (27 kB)
Collecting virtualenv==20.0.20
  Using cached virtualenv-20.0.20-py2.py3-none-any.whl (4.7 MB)
Collecting vmc-client-bindings==1.32.0
  Using cached https://cdn.githubraw.com/saltstack/vsphere-automation-sdk-python/master/lib/vmc-client-bindings/vmc_client_bindings-1.32.0-py2.py3-none-any.whl
Collecting vmc-draas-client-bindings==1.17.0
  Using cached https://cdn.githubraw.com/saltstack/vsphere-automation-sdk-python/master/lib/vmc-draas-client-bindings/vmc_draas_client_bindings-1.17.0-py2.py3-none-any.whl
Collecting vsphere-automation-sdk==1.46.0
  Using cached https://cdn.githubraw.com/saltstack/vsphere-automation-sdk-python/master/lib/vsphere-automation-sdk/vSphere_Automation_SDK-1.46.0-py3-none-any.whl
Collecting watchdog==0.9.0
  Using cached watchdog-0.9.0-py3-none-any.whl
Collecting websocket-client==0.40.0
  Using cached websocket_client-0.40.0-py2.py3-none-any.whl
Collecting werkzeug==0.15.6
  Using cached Werkzeug-0.15.6-py2.py3-none-any.whl (328 kB)
Collecting wrapt==1.11.1
  Using cached wrapt-1.11.1-cp39-cp39-linux_x86_64.whl
Collecting xmltodict==0.12.0
  Using cached xmltodict-0.12.0-py2.py3-none-any.whl (9.2 kB)
Requirement already satisfied: yamlordereddictloader==0.4.0 in ./.nox/pytest-parametrized-3-crypto-none-transport-zeromq-coverage-false/lib/python3.9/site-packages (from -r requirements/static/ci/py3.9/linux.txt (line 248)) (0.4.0)
Collecting zc.lockfile==1.4
  Using cached zc.lockfile-1.4-py3-none-any.whl
Collecting zipp==3.4.0
  Using cached zipp-3.4.0-py3-none-any.whl (5.2 kB)
Requirement already satisfied: setuptools in ./.nox/pytest-parametrized-3-crypto-none-transport-zeromq-coverage-false/lib/python3.9/site-packages (from jsonschema==3.2.0->-r requirements/static/ci/py3.9/linux.txt (line 149)) (54.2.0)
WARNING: Installing from a newer Wheel-Version (1.1)
Installing collected packages: wrapt, websocket-client, vapi-common-client, transitions, tempora, smmap, scp, python-jose, pytest, pyserial, pyasn1-modules, passlib, oscrypto, ntc-templates, networkx, netaddr, ncclient, junit-xml, jsonpickle, jsonpatch, jinja2, docker-pycreds, dnspython, colorama, click, cachetools, backports.functools-lru-cache, azure-storage-common, azure-mgmt-web, azure-mgmt-trafficmanager, azure-mgmt-subscription, azure-mgmt-storage, azure-mgmt-sql, azure-mgmt-signalr, azure-mgmt-servicefabric, azure-mgmt-servicebus, azure-mgmt-search, azure-mgmt-scheduler, azure-mgmt-resource, azure-mgmt-reservations, azure-mgmt-relay, azure-mgmt-redis, azure-mgmt-recoveryservicesbackup, azure-mgmt-recoveryservices, azure-mgmt-rdbms, azure-mgmt-powerbiembedded, azure-mgmt-policyinsights, azure-mgmt-notificationhubs, azure-mgmt-network, azure-mgmt-msi, azure-mgmt-monitor, azure-mgmt-media, azure-mgmt-marketplaceordering, azure-mgmt-maps, azure-mgmt-managementpartner, azure-mgmt-managementgroups, azure-mgmt-machinelearningcompute, azure-mgmt-logic, azure-mgmt-loganalytics, azure-mgmt-keyvault, azure-mgmt-iothubprovisioningservices, azure-mgmt-iothub, azure-mgmt-iotcentral, azure-mgmt-hanaonazure, azure-mgmt-eventhub, azure-mgmt-eventgrid, azure-mgmt-dns, azure-mgmt-devtestlabs, azure-mgmt-devspaces, azure-mgmt-datamigration, azure-mgmt-datalake-store, azure-mgmt-datalake-analytics, azure-mgmt-datafactory, azure-mgmt-cosmosdb, azure-mgmt-containerservice, azure-mgmt-containerregistry, azure-mgmt-containerinstance, azure-mgmt-consumption, azure-mgmt-compute, azure-mgmt-commerce, azure-mgmt-cognitiveservices, azure-mgmt-cdn, azure-mgmt-billing, azure-mgmt-batchai, azure-mgmt-batch, azure-mgmt-authorization, azure-mgmt-applicationinsights, azure-mgmt-advisor, azure-cosmosdb-nspkg, aws-sam-translator, zipp, zc.lockfile, xmltodict, werkzeug, vmc-draas-client-bindings, vmc-client-bindings, vapi-client-bindings, suds-jurko, sshpubkeys, responses, pyzmq, pyvmomi, pytest-tempdir, pyeapi, psutil, portend, pathtools, nsx-vmc-policy-python-sdk, nsx-vmc-aws-integration-python-sdk, nsx-python-sdk, nsx-policy-python-sdk, netmiko, msgpack, mock, junos-eznc, jsondiff, ipaddress, google-auth, gitdb, geomet, filelock, docker, distlib, contextlib2, ciscoconfparse, cheroot, cfn-lint, certvalidator, cached-property, boto, azure-storage-queue, azure-storage-file, azure-storage-blob, azure-servicemanagement-legacy, azure-servicefabric, azure-servicebus, azure-mgmt, azure-loganalytics, azure-keyvault, azure-graphrbac, azure-eventgrid, azure-datalake-store, azure-cosmosdb-table, azure-batch, azure-applicationinsights, aws-xray-sdk, argh, appdirs, watchdog, vsphere-automation-sdk, virtualenv, vcert, timelib, strict-rfc3339, sqlparse, setproctitle, rfc3987, python-gnupg, python-etcd, pytest-subtests, pytest-salt-factories, pytest-helpers-namespace, pyinotify, pyiface, pygit2, pycurl, pycryptodomex, napalm, moto, mako, libnacl, kubernetes, keyring, kazoo, jxmlease, gitpython, genshi, flaky, distro, croniter, clustershell, cherrypy, cheetah3, cassandra-driver, azure, apache-libcloud
  Attempting uninstall: wrapt
    Found existing installation: wrapt 1.12.1
ERROR: Cannot uninstall 'wrapt'. It is a distutils installed project and thus we cannot accurately determine which files belong to it which would lead to only a partial uninstall.
nox > Session pytest-parametrized-3(crypto=None, transport='zeromq', coverage=False) failed.
nox > Ran multiple sessions:
nox > * pytest-zeromq-3(coverage=False): success
nox > * pytest-parametrized-3(crypto=None, transport='zeromq', coverage=False): failed
```

Is this a known bug or am I missing something?

Ch3LL commented:
I have not seen that before. Can you share the nox command you are running and the output from `pip list |grep -i nox`

dithmer commented:
> pip list |grep -i nox

nox                                   2020.12.31

dithmer committed:
Merge branch 'master' into fix-hetzner-saltcloud-driver-privatenet

Ch3LL commented:
and what is the entire nox command you are running, and python version?

dithmer commented:
Python version: 3.9.2

Full command `nox -e 'pytest-zeromq-3(coverage=False)' -- tests/pytests/unit/cloud/clouds/test_hetzner.py`

Ch3LL commented:
hmm I'm still not able to replicate it. ping @s0undt3ch do you have any ideas for hte above error ^

s0undt3ch commented:
@dithmer please try deleting the `.nox` directory and running again.

Commit:
Change naming according to hetzner doc

Commit:
Fix getting the right attribuyte vom hetzner private net

Commit:
Test public_ip and private_net convert to dict functions

dithmer commented:
Seems like `wrapt` is copied by `nox` from the global python installation and not from the actual activated virtual environment. If this is installed as a dist-package and not as an site-package it is not partially uninstallable. The attempt to install a different version by nox then, can't work. Fixed it at my site, but that's definitely a weird behaviour of nox.

Managed to get the tests running and added some assertions for the convert methods from public and private nets. Anything else to do here @Ch3LL ?

dithmer committed:
Merge branch 'master' into fix-hetzner-saltcloud-driver-privatenet

dithmer committed:
Merge branch 'master' into fix-hetzner-saltcloud-driver-privatenet
